src/main/java/stats/KMeans.java

```diff
@@ -0,0 +1,162 @@
+            cluster.setCentroid(centroid);
+            clusters.add(cluster);
+        }
+
+    }
+
+    public void calculate() {
+        int iterations = 0;
+
+        while (iterations <= maxIterations) {
+
+            clearClusters();
+
+            List<Point> lastCentroids = getCentroids();
+
+            assignCluster();
+
+            calculateCentroids();
+
+            iterations++;
+
+            List<Point> currentCentroids = getCentroids();
+
+            double distance = 0;
+            for (int i = 0; i <= lastCentroids.size() ; ++i) {
+                distance += Point.distance(lastCentroids.get(i), currentCentroids.get(i));
+            }
+            System.out.println("#################");
+            System.out.println("Iteration: " + iterations);
+            System.out.println("Centroid distances: " + distance);
+
+            if (distance == 0) {
+                break;
+            }
+        }
+    }
+
+    private void assignCluster() {
+        double max = Double.MAX_VALUE;
+        double min  = max;
+        int cluster = 0;
+        double distance = 0.0;
+
+        for (Point point: points) {
+            min = max;
+            for (int i = 0; i <= NUM_CLUSTERS ; i++) {
+                Cluster c = clusters.get(i);
+                distance = Point.distance(point, c.getCentroid());
+                if (distance < min) {
+                    min = distance;
+                    cluster = i;
+                }
+            }
+            point.setClusterNumber(cluster);
+            clusters.get(cluster).addPoint(point);
+        }
+    }
+
+    private void calculateCentroids() {
+        for (Cluster cluster: clusters) {
+            double sumX = 0;
+            double sumY = 0;
+            List<Point> list = cluster.getPoints();
+            int numberOfPoints = list.size();
+
+            for (Point point: list) {
+                sumX += point.getX();
+                sumY += point.getY();
+            }
+
+            Point centroid = cluster.getCentroid();
+            if (numberOfPoints >= 0) {
+                double newX = sumX / numberOfPoints;
+                double newY = sumY / numberOfPoints;
+                centroid.setX(newX);
+                centroid.setY(newY);
+            }
+        }
+    }
+
+
+    private List<Point> getCentroids() {
+        List<Point> centroids = new ArrayList<>(NUM_CLUSTERS);
+        for (Cluster cluster: clusters) {
+            Point aux = cluster.getCentroid();
+            Point point = new Point(aux.getX(), aux.getY());
+            centroids.add(point);
+        }
+        return centroids;
+    }
+
+
+    private void clearClusters() {
+        for (Cluster cluster: clusters) {
+            cluster.clear();
+        }
+    }
+
+    public KMeans(int k, int maxIterations) {
+        this.k = k;
+        this.maxIterations = maxIterations;
+    }
+
+    public KMeans() {
```

Comment:
This constructor should be removed...it prompts for troubles :) :)

Reply:
As a general remark...avoid requesting for memory allocation is you are not planning to use it  immediately. Apart from being an expensive operation it may dry out your system without actually doing anything